### DISCUSSION TITLE
Issue in Airflow Downgrading from v2.7.0 to lower versions v2.6.0/v2.3.4

### DISCUSSION THREAD

**Discussion by github-aman** [apache/airflow#36901] [Q&A]
Hello All,

Could you please help me on downgrading Airflow setup running on ECS service as i am getting below error:-

You need to upgrade the database. Please run airflow db upgrade. Make sure the command is run using Airflow version 2.6.0

I am at currently Airflow v2.7.0 but need to perform downgrade just to perform rollback testing if something breaks in my app.

I have performed already below in the script which kicks off the Airflow from script.

airflow db upgrade
airflow db downgrade --from-version 2.7.0 --to-version 2.6.0
airflow db migrate

Please suggest on above !!

**Comment by potiuk** [comment]
Likely you have some mistake into where and how you run it - you do not write in which circumstances, where, which compoent you get the error. If you want help you need to describe it more precisely. what you run, what deployment, where you get the error, actual logs that you get etc. etc. Likely why gathering it, you will find the reason.

**Reply by github-aman** [reply]
**Scenario:-**

I was having working test environment v2.7.0 Airflow setup which i had upgraded from v2.6.0. Now I am trying to downgrade it back to v2.6.0 or further till v2.3.0.

**Process:-**

- I am using dockerfile which is building the image & pushes to ECR which will further used by AWS ECS service.
- I restart the ECS tasks which takes up the latest images & spins it & setup comes online or become accessible.
- I am using one entrypoint.sh script which executes in the last step of dockerfile so that as soon as the AWS ECS tasks consumes the image from ECR, first thing is the entrypoint script execution which is having below set of commands(attachment):-

![entrpoint_script](https://github.com/apache/airflow/assets/32449553/8c4fc702-9847-450a-b584-9499f3ec5c2b)

Let me know if any input !!

**Reply by github-aman** [reply]
Attaching two more error logs from AWS ECS webserver tasks.
![webserver_errorlogs](https://github.com/apache/airflow/assets/32449553/9b1f4202-191f-4d47-ace5-78b096ee71d7)
![webserver_errorlogs_2](https://github.com/apache/airflow/assets/32449553/5d294e4f-e1f7-4577-b7bc-7fe7875b1c45)

**Reply by potiuk** [reply]
You seem to be running downgrade using airflow 2.6.0. you cannot downgrade from airflow 2.7.0 to 2.6.0  while running 2.6.0. Airflow 2.6.0 knows nothing about airflow 2.7.0 - so there is no way it can downgrade from it. The downgrade you need to do MUST be down using airflow 2.7.0 (after shutting down running airflow - ideally in a separate job. Only after that you should switch your images to 2.6.0

**Reply by github-aman** [reply]
No, i am downgrading from v2.7.0 only to lower versions v2.3.4 & getting below error:-

- Please run `airflow db upgrade`. Make sure the command is run using Airflow version 2.3.4.

Could you help me ? if there is direct command where i can pass version as well with command "airflow db upgrade".


I had tested below scenario as well where i got such errors:-

1. Downgrade from v2.7.0 to v2.5.0 | error attached.
![libpq_error](https://github.com/apache/airflow/assets/32449553/2b40309f-1a46-4314-afe5-1da356ac606e)

**Reply by potiuk** [reply]
This is the same story. You are running your `airflow db upgrade` using 2.3.4 but your DB is 2.7.0. Make sure you downgrade your DB before you downgrade the DB. It's not going to work otherwise.

**Reply by github-aman** [reply]
I got your point.
Just want to understand from you, how should i proceed with this db downgrade from script point of view? any specific command which you can suggest, like in case of up-gradation i use ( airflow db upgrade ) & it helps me, but here things are not working programmatically or from docker image & entrypoint.sh script side.

Just to update you below as well:-

I was going through official documentation of airflow & found these alembic versions which represents every version uniquely, so i ran the query to change the alembic version directly through sql query & changed the version from v2.7.0 to v2.3.4 & restarted all the AWS ECS services & it worked fine & setup finally came up at v2.3.4, which i wanted.

Is it the right way? 
Should i be covering everything if i change alembic version using SQL query? or command performs some other upgrade/downgrade as well(airflow db upgrade)?

**Reply by potiuk** [reply]
> I was going through official documentation of airflow & found these alembic versions which represents every version uniquely, so i ran the query to change the alembic version directly through sql query & changed the version from v2.7.0 to v2.3.4 & restarted all the AWS ECS services & it worked fine & setup finally came up at v2.3.4, which i wanted.

No specific suggestions - it's up to yoy, some people use ansible, some people do it manually, som develop their own scripts. We do not want to be opinionated here - we just give you the tools that you can put together as you wish. I'd personally likely simply have several images with different versions of airflow and use docker to run bash scripts mounted to these docker files, or - have several virtualenvs installed independently and rather than downgrading the virtualenv treat them as immutable - once installed, never upgrade/downgrade airflow - I 'd name the venvs like `ariflow-2.8.1`, `airflow-2.8.2` etc. and have my own scripts/ansible/whatever to be able to create/use such docker images or venvs as needed in reproducible way.

> I was going through official documentation of airflow & found these alembic versions which represents every version uniquely, so i ran the query to change the alembic version directly through sql query & changed the version from v2.7.0 to v2.3.4 & restarted all the AWS ECS services & it worked fine & setup finally came up at v2.3.4, which i wanted.

No it's very vrong. The migrate command not only updates the version in the table but also runs migration scripts that are embedded in Airflow package. That's why you cannot downgrade from 2.6.0 to 2.3.4 by using Airflow 2.3.4, because Airflow 2.3.4 does not have the migration scripts that are applied by alembic to migrate down from 2.6.0 - see https://github.com/apache/airflow/tree/main/airflow/migrations/versions for all  the scripts (each script has up/down steps and alembic executes them when doing up/down grading.

If you manually modified those versions, you have generally broken the relation between the structure and the recorded version, so your database is generally broken. My recommendation is you wipe it all and start from scratch by following the above,

**Reply by github-aman** [reply]
- It's up to you, some people use ansible, some people do it manually, som develop their own scripts. 

**Response:-** Can you tell me the direct command if same i can use in my bash script? i am looking for exact thing which i can pass similar to "airflow db upgrade" or anything which will work?



**Below is for my understanding on multiple virtual env's as below:-**

-  or - have several virtualenvs installed independently and rather than downgrading the virtualenv treat them as immutable - once installed, never upgrade/downgrade airflow - I 'd name the venvs like ariflow-2.8.1, airflow-2.8.2 etc.
**Response:-** When you say above multiple virtualenvs & multiple airflow versions/env's, Is  my understanding correct that you have multiple db connections for different airflow versions/env's, in that case it can be immutable | Agreed.



- No it's very wrong. The migrate command not only updates the version in the table but also runs migration scripts that are embedded in Airflow package. That's why you cannot downgrade from 2.6.0 to 2.3.4 by using Airflow 2.3.4, because Airflow 2.3.4 does not have the migration scripts that are applied by alembic to migrate down from 2.6.0
**Response:-** Got your point, but still i am looking out a way to run this script on some particular version so that i can downgrade if something does not proceed as per my expectation.



- My recommendation is you wipe it all and start from scratch by following the above.
Response:- I can't setup from scratch. it's going to act as fresh installation/new setup. i have 200+ users 80+ custom roles, multiple dags in EFS, need to preserve the metadata as well, these are production systems.



**Your direct command response which i can simply pass in my bash script is highly appreciated !!**

**Reply by potiuk** [reply]
I think you have all the information now to proceed. I can't do the job for you.

**Reply by potiuk** [reply]
Read the docs, experiment. implement. It's all there (and you got enough leads in the above comments). This is not scalable to consult every single person who uses airflow (there are tens of thousands of them) and create their own deployment scripts for them. You are paid to do the job and got enough help to do so.

**Reply by github-aman** [reply]
Thank You for the quick support & help !!
